chapter_explainability/global_model_analysis.md

```diff
@@ -0,0 +1,305 @@
+                    x[:,feat_idx_1] = u_1
+                temp_1.extend(torch.sigmoid(self.model(x)).detach() \
+                              .numpy()[:, target])
+            temp_0.append(sum(temp_1) / len(temp_1))
+        Y_hat.append(temp_0) if len(features) > 1 else Y_hat.extend(temp_0)
+    d2l.plt.figure(figsize=(2, 2))
+    if len(features) > 1: 
+        X, Y = np.meshgrid(unique_val_1, unique_val_0)
+        c = d2l.plt.contourf(X, Y, np.array(Y_hat), cmap='GnBu')
+        d2l.plt.xlabel(features[1])
+        d2l.plt.ylabel(features[0])
+        d2l.plt.legend(c.legend_elements()[0], [f'{a:.2f}' for a in c.levels], 
+                       loc=(1.05, 0), fontsize='9')
+    else:
+        d2l.plt.xlabel(features[0])
+        d2l.plt.ylabel('Probability of target=' + str(target))
+        if feature_type.lower() in ['cat', 'categorical']:
+            d2l.plt.xticks(unique_val_0)
+            d2l.plt.bar(unique_val_0, Y_hat) 
+        else:
+            d2l.plt.plot(unique_val_0, Y_hat)
+```
+
+The following partial dependence plot shows how the predicted probability of target=1 (on the vertical axis) changes as we increase the cholesterol level (on the horizontal axis).
+
+```{.python .input  n=15}
+explainer.PartialDependencePlot(['chol'], 1, feature_type='NUMERICAL')
+```
+
+PDPs with two input features of interest show the interactions among the two features. The PDP below shows how the predicted probability of target=1 varies with the changes of *thalach* (maximum heart rate achieved) and *age*.
+
+```{.python .input  n=16}
+explainer.PartialDependencePlot(['thalach', 'age'], 1)
+```
+
+PDPs are an intuitive way to extract insights from black-box models. Disadvantages of PDPs are: (1) Due to the limits of human perception, we can only show at most three features of interest at a time.  (2) PDP assumes that the features of interest $X_T$ are independent from other features $X_S$, but this assumption is often violated in practice. In the case of correlated features, we will create absurd or unrealistic data points when computing PDP, which may cause mis-interpretations. (3) Another pitfall in PDP is that it does not show the feature distribution, we may overinterpret some regions with very few data points (e.g., *chol*>400).
+
+## Summary
+* Error analysis helps us get insights into the erroneous cases.
+* Feature importance can be obtained via ablation studies.
+* Partial dependence plots can show the relationship between targets and features of interest.
+
+## Exercises
+1. Analyze the error distribution across different cholesterol levels. 
+2. Can you calculate the importance score of the batch norm via ablation study? Hint: You may need to revise the MLP model.
+3. Compute the global feature importance using *REMOVE* mode and compare the results and runtime with the *PERMUTE* mode. What can you find from the comparison?
+4. Analyze the relationship between *target* and resting blood pressure.
```

Comment:
2->1, 3->1, 4->1

Reply:
Thanks but I don't get this comment.

Reply:
For numbered bullet points we can simply use:
```
1. Analyze ...
1. Can you ...
1. Compute ...
1. Analyze ...
```

It will be easier to re-order them or insert/delete bullets without the need for re-numbering

Reply:
I see, thanks for the suggestion!